File: iOS-POC-2/Coordinators/TodoFlowCoordinator.m
#import "TodoFlowCoordinator.h"
#import "CompletionViewController.h"
@import Feature;

@interface TodoFlowCoordinator ()
@property (nonatomic, weak) UINavigationController *navigationController;
@property (nonatomic, weak) TodoInputViewController *inputViewController;
@end

@implementation TodoFlowCoordinator

- (instancetype)initWithNavigationController:(UINavigationController *)navigationController
                         inputViewController:(TodoInputViewController *)inputViewController {
    self = [super init];
    if (self) {
        _navigationController = navigationController;
        _inputViewController = inputViewController;
        inputViewController.flowDelegate = self;
    }
    return self;
}

#pragma mark - TodoInputViewControllerDelegate

- (void)todoInputViewController:(TodoInputViewController *)controller
                  didSubmitText:(NSString *)text {
    __weak typeof(self) weakSelf = self;
    // 確認フロー（確認画面1→2）は Swift の Feature パッケージが提供する。
    UIViewController *flow = [ConfirmFlowFactory
        makeConfirmFlowWithText:text
                    onCompleted:^{ [weakSelf showCompletion]; }
                         onExit:^{ [weakSelf.navigationController popViewControllerAnimated:YES]; }];
    [self.navigationController pushViewController:flow animated:YES];
}

#pragma mark - Navigation

- (void)showCompletion {
    // 保存完了 → 入力欄をリセットし、完了画面へ。
    [self.inputViewController resetInput];

    __weak typeof(self) weakSelf = self;
    CompletionViewController *completionVC = [[CompletionViewController alloc] init];
    completionVC.onBackToInput = ^{
        [weakSelf.navigationController popToRootViewControllerAnimated:YES];
    };
    [self.navigationController pushViewController:completionVC animated:YES];
}

@end

File: iOS-POC-2/Coordinators/TodoFlowCoordinator.swift
import UIKit
import Feature

/// Todo の入力〜確認〜完了フローの画面遷移を一元管理する Coordinator。
/// 各画面（入力・完了）は自身の遷移を知らず、この Coordinator が順序を所有する。
/// ObjC 画面（TodoInputViewController / CompletionViewController）を扱うためアプリターゲットに置く。
@MainActor
@objc final class TodoFlowCoordinator: NSObject, TodoInputViewControllerDelegate {
    private weak var navigationController: UINavigationController?
    private weak var inputViewController: TodoInputViewController?

    @objc init(navigationController: UINavigationController,
               inputViewController: TodoInputViewController) {
        self.navigationController = navigationController
        self.inputViewController = inputViewController
        super.init()
        inputViewController.flowDelegate = self
    }

    // MARK: - TodoInputViewControllerDelegate

    func todoInputViewController(_ controller: TodoInputViewController, didSubmitText text: String) {
        // 確認フロー（確認画面1→2）は Swift の Feature パッケージが提供する。
        let flow = ConfirmFlowFactory.makeConfirmFlow(
            text: text,
            onCompleted: { [weak self] in self?.showCompletion() },
            onExit: { [weak self] in self?.navigationController?.popViewController(animated: true) }
        )
        navigationController?.pushViewController(flow, animated: true)
    }

    // MARK: - Navigation

    private func showCompletion() {
        // 保存完了 → 入力欄をリセットし、完了画面へ。
        inputViewController?.resetInput()

        let completionVC = CompletionViewController()
        completionVC.onBackToInput = { [weak self] in
            self?.navigationController?.popToRootViewController(animated: true)
        }
        navigationController?.pushViewController(completionVC, animated: true)
    }
}
